src/libraries/System.Text.Json/docs/ReferenceHandling_spec.md

```diff
@@ -0,0 +1,810 @@
+  - [In depth](#in-depth)
+  - [Compatibility](#compatibility)
+- [Examples](#examples)
+  - [Using Default on Serialize](#using-default-on-serialize)
+  - [Using Ignore on Serialize](#using-ignore-on-serialize)
+  - [Using Preserve on Serialize](#using-preserve-on-serialize)
+  - [Using Preserve on Deserialize](#using-preserve-on-deserialize)
+- [Other languages](#other-languages)
+  - [Newtonsoft.Json](#newtonsoftjson)
+  - [dojo toolkit (JavaScript framework)](#dojo-toolkit-javascript-framework)
+  - [flatted (JavaScript module) (probably not worth it)](#flatted-javascript-module-probably-not-worth-it)
+  - [Jackson (Java)](#jackson-java)
+  - [golang](#golang)
+- [Ground rules](#ground-rules)
+  - [Reference objects ($ref)](#reference-objects-ref)
+  - [Preserved objects ($id)](#preserved-objects-id)
+  - [Preserved arrays](#preserved-arrays)
+  - [JSON Objects if not Collection (Class | Struct | Dictionary) - On Deserialize (and Serialize?)](#json-objects-if-not-collection-class--struct--dictionary---on-deserialize-and-serialize)
+  - [JSON Object if Collection - On Deserialize](#json-object-if-collection---on-deserialize)
+  - [Immutable types](#immutable-types)
+  - [Value types](#value-types)
+  - [Interaction with JsonPropertyNameAttribute](#interaction-with-jsonpropertynameattribute)
+- [Future](#future)
+- [Notes](#notes)
+
+
+# Terminology
+
+**Reference loops**: also referred as circular references, occur when a property of the object refers to the object itself, directly (a -> a) or indirectly (a -> b -> a). They also occur when the element of an array refers to the array itself (arr[0] -> arr). Multiple occurrences of the same reference does not imply circularity.
+
+**Preserve duplicated references**: Semantically represent objects and/or arrays whose have been previously written, with a reference to them in subsequent founds.
+
+**Metadata**: Extra properties on JSON objects and/or arrays (that may change their schema) to enable reference preservation when round-tripping, those properties are only meant to be understand by the `JsonSerializer`.  
+
+# Motivation
+
+Currently there is no mechanism to prevent infinite looping in circular objects (while serializing) nor to preserve references that round-trip when using System.Text.Json.
+
+This is a heavily requested feature since it is consider by many as a very common scenario, specially when serializing POCOs that came from an ORM Framework, such as Entity Framework; even though the JSON specification does not support reference loops by default. Therefore, this will be implemented as an opt-in feature (for both serialization and deserialization).
+
+The current solution to deal with reference loops is to rely in MaxDepth and throw a JsonException after it is exceeded. Now, this is a decent and cheap solution but we will also offer other not-so-cheap options to deal with this problem while keeping the current one in order to not affect the out-of-the-box performance.
```

Comment:
```suggestion
The current solution to deal with cycles in the object graph while serializing is to rely on `MaxDepth` and throw a `JsonException` after it is exceeded. This was done to avoid perf overhead for cycle detection in the common case. The goal is to enable the new opt-in feature with minimal impact to existing performance.

```